Modules/_json.c

```diff
@@ -1503,6 +1503,13 @@ encoder_listencode_obj(PyEncoderObject *s, _PyAccu *acc,
             }
         }
         newobj = PyObject_CallFunctionObjArgs(s->defaultfn, obj, NULL);
+        if (obj != NULL && newobj != NULL && newobj->ob_type == obj->ob_type) {
```

Comment:
How about `if (PyObject_IsInstance(newobj, (PyObject *) Py_TYPE(obj))) {` ?

Also, I think this can put above the check of `if (newobj == NULL)`